design-docs/005-manifest-split.md

```diff
@@ -0,0 +1,406 @@
+
+## Discussion
+
+This an optimization problem and there are many ways to decide what "ideal"
+means. Examples:
+
+* Datasets that are periodically appended along the time dimension, may
+benefit from putting in the same manifest file several arrays split by time.
+Most reads and write will only need to look at the latest manifest.
+* Small arrays, particularly coordinate arrays, would benefit from packing
+multiple arrays in a single, small manifest file that can be read quickly during
+interactive usage.
+* Huge arrays require splitting into multiple manifest files, and it becomes
+important to know along what dimension they should be split for effective
+locality.
+
+At this point, we want to build only the first iteration of this. We don't need
+to be super smart, as long as interactive usage is not too bad. Developing the
+first iteration would help better defining the datastructures and changes
+required to the on-disk format.
+
+Since there are so many cases, we want the ability to let the user configure
+the algorithm to some degree. That way we don't need to be as smart with the
+defaults, while at the same time offering full power for advanced users.
+
+We have analyzed real world datasets, and verified that there is a bimodal
+distribution of number of chunks per array. Some arrays have very low number
+of chunks, these are usually the coordinate arrays; and some others have a much larger
+number of chunks, the actual variables. We can use this fact to pack together the
+manifests for all small arrays, in a small, fast to parse manifest. This is
+probably a good default behavior, if we put it together with a limit to
+the manifest sizes for the remaining arrays.
+
+## Design
+
+We don't implement array splitting at this time. Each array will go fully into
+a single manifest. This is of course a limitation for very large arrays, but one
+we will lift in a future iteration. The on-disk format model supports the split
+and the current design must not block it.
+
+We define _manifest sets_. A manifest set is a group of individual array manifests that
+are stored together allowing a user to optimize for certain types of arrays. We let the user define
+the sets, but we offer good defaults that should work reasonably well in
+common cases.
+
+We let the user indicate _rules_ that assign arrays to
+a specific manifest set. These rules can be based on array paths or number
+of chunks. In the future we could add more power to the rules system.
+
+Other new feature we provide is the ability to preload certain manifests.
+These manifests will be fetched in the background as soon as a session is
+created, speeding up interactive use. In the future, we may add the ability
+to prefetch manifests for certain branches/tags as soon as the Repository is
+open.
+
+Manifest sets, array rules, and manifest prefetch, are configured in the
+persistent configuration of the repository and can be overloaded on open as
+any other configuration value.
```

Comment:
How is the default configuration provided?

Reply:
And what happens if the _configuration is changed_? Will all of the manifests be rewritten?

Reply:
I'll explain in the document, i missed this part. No, we won't rewrite until they are rewritten by a commit. In the future we could offer repacking the manifest as a explicit optimization operation.